{{cookiecutter.project_slug}}/backend/app/app/crud/base.py

```diff
@@ -0,0 +1,161 @@
+
+            # model definition in app/db_models/item.py
+            class Item(Base):
+                id: int
+                ...
+
+            # crud definition in app/crud/item.py
+            from app.db_models.item import Item
+            from app.models.item import ItemUpdate, ItemCreate
+            from app.crud.base import CrudBase
+
+
+            class CrudItem(CrudBase):
+
+                def get(self, db_session: Session, obj_id: int) -> Optional[Item]:
+                    return super(CrudItem, self).get(db_session, obj_id=obj_id)
+
+                def get_multi(self, db_session: Session, *, skip=0, limit=100) -> List[Optional[Item]]:
+                    return super(CrudItem, self).get_multi(db_session, skip=skip, limit=limit)
+
+                def create(self, db_session: Session, *, obj_in: ItemCreate) -> Item:
+                    return super(CrudItem, self).create(db_session, obj_in=obj_in)
+
+                def update(self, db_session: Session, *, obj: Base, obj_in: ItemUpdate) -> Item:
+                    return super(CrudItem, self).update(db_session, obj=obj, obj_in=obj_in)
+
+
+            crud_item = CrudItem(Item)
+
+        Arguments:
+            db_model {Base} -- Class of the DB model which CRUD methods will be provided for
+        """  # noqa
+        self.db_model = db_model
+
+    def get(self, db_session: Session, obj_id: int) -> Optional[Base]:
+        """
+        get returns the object from the Database that matches the given obj_id
+        
+        Arguments:
+            db_session {Session} -- Dependency injection of the Database session, which will be used to commit/rollback changes.
+            obj_id {int} -- ID of the object in the Database. It must be defined by a PrimaryKey on the 'id' column.
+        
+        Returns:
+            Optional[Base] -- Returns an instance of self.db_model class if an object is found in the Database for the given obj_id. Returns None if there is no match found.
+        """  # noqa
+        return db_session.query(self.db_model).get(obj_id)
+
+    def get_multi(self, db_session: Session, *, skip=0, limit=100) -> List[Optional[Base]]:
+        """
+        get_multi queries all Database rows, without any filters, but with offset and limit options (for pagination purpose)
+        
+        Arguments:
+            db_session {Session} -- Dependency injection of the Database session, which will be used to commit/rollback changes.
+        
+        Keyword Arguments:
+            skip {int} -- Number of rows to skip from the results (default: {0})
+            limit {int} -- Maximum number of rows to return (default: {100})
+        
+        Returns:
+            List[Optional[Base]] -- Array of DB instances according given parameters. Might be empty if no objets are found.
+        """  # noqa
+        return db_session.query(self.db_model).offset(skip).limit(limit).all()
+
+    def create(self, db_session: Session, *, obj_in: BaseModel) -> Base:
```

Comment:
After using this, I noticed you could make it more general by passing in extra arguments for db_model.

```py
def create(self, db_session: Session, *, obj_in: BaseModel, **kwargs) -> Base:
    obj_in_data = jsonable_encoder(obj_in)
    obj = self.db_model(**obj_in_data, **kwargs)
    db_session.add(obj)
    db_session.commit()
    db_session.refresh(obj)
    return obj
```

Commenly parent_id, child_id, and similar database only args would go in there.

Reply:
I think any other attributes you might want to pass at creation time must be defined in your schema, and will be passed through `obj_in.

That is also the case with relationships like parent_id and child_id.

Let's say your User belongs to a city. The schemas would look like this:

```python
# Shared properties
class UserBase(BaseModel):
    email: str = None
    is_active: bool = True
    is_superuser: bool = False
    full_name: str = None
    city_id: int = None

    class Config:
        orm_mode = True


# Additional properties stored in DB
class UserBaseInDB(UserBase):
    id: int = None
    created_at: Optional[datetime] = None
    updated_at: Optional[datetime] = None


# Properties to receive via API on creation
class UserCreate(UserBaseInDB):
    email: str
    password: str
    created_at: datetime = Schema(datetime.utcnow())


# Properties to receive via API on update
class UserUpdate(UserBaseInDB):
    password: str = None
    updated_at: datetime = Schema(datetime.utcnow())
```

Reply:
Oh I see, that makes sense thanks.